### PR TITLE
Add gogo benchmark

### PR DESCRIPTION
**Pull request by BSBandme**

### CODE REVIEW

File: benchmarks/Makefile.am [hunk 1]
```diff
@@ -253,50 +252,313 @@ benchmarks_protoc_inputs_proto2_message4 =                                 \
...
+make_tmp_dir_gogo:
+	mkdir -p tmp/go_no_group
+	mkdir -p tmp/gogofast
+	mkdir -p tmp/gogofaster
+	mkdir -p tmp/gogoslick
+	touch make_tmp_dir_gogo
+
+if USE_EXTERNAL_PROTOC
+
+go_no_group_protoc_middleman: make_tmp_dir_gogo gogo_proto_middleman
```
Comment (BSBandme) [inline benchmarks/Makefile.am]: Still need to list all protoc rules for different gogo, since different gogo are different plugins.

Comment (BSBandme) [reply benchmarks/Makefile.am]: Any updates on this?

**Comment by BSBandme** [comment]
Any updates on this?

File: benchmarks/Makefile.am (hunk 1 above)
Comment (xfxyjwf) [reply benchmarks/Makefile.am]: Why can't it be parameterized?

File: benchmarks/Makefile.am [hunk 2]
```diff
@@ -318,7 +580,26 @@ CLEANFILES =                                                               \
 	python-cpp-generated-code-benchmark                                      \
 	go-benchmark                                                             \
 	go_protoc_middleman                                                      \
-	go_protoc_middleman2
+  make_tmp_dir_gogo                                                        \
```
Comment (xfxyjwf) [inline benchmarks/Makefile.am]: Fix all the incorrect alignments.

File: benchmarks/gogo/cpp_no_group/cpp_benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,253 @@
+// Protocol Buffers - Google's data interchange format
```
Comment (xfxyjwf) [inline benchmarks/gogo/cpp_no_group/cpp_benchmark.cc]: Why can't we reuse the old code?

File: benchmarks/gogo/go_no_group/go_benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
+package main
```
Comment (xfxyjwf) [inline benchmarks/gogo/go_no_group/go_benchmark_test.go]: Can we use a script to create these variants from a common template?

File: benchmarks/util/data_proto2_to_proto3_util.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+using google::protobuf::FieldDescriptor;
+using google::protobuf::Message;
+using google::protobuf::Reflection;
+
+namespace google {
+namespace protobuf {
+namespace util {
+
+class DataGroupStripper {
```
Comment (xfxyjwf) [inline benchmarks/util/data_proto2_to_proto3_util.h]: Why not move this to gogo_data_scrubber.cc?

File: benchmarks/util/gogoproto_generator.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
+// TODO: Insert description here. (generated by yilunchong)
```
Comment (xfxyjwf) [inline benchmarks/util/gogoproto_generator.h]: Merge this file with protoc-gen-gogoproto.cc

File: benchmarks/util/gogo_data_scrubber.cc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#include "util/data_proto2_to_proto3_util.h"
+
+#include <fstream>
+
+using google::protobuf::util::DataGroupStripper;
+
+std::string ReadFile(const std::string& name) {
+  std::ifstream file(name.c_str());
+  GOOGLE_CHECK(file.is_open()) << "Couldn't find file '" << name <<
+                                  "', please make sure you are running "
```
Comment (xfxyjwf) [inline benchmarks/util/gogo_data_scrubber.cc]: The format should be:
```cpp
GOOGLE_CHECK(...) << "xxxx"
                  << "xxx";

```

File: benchmarks/util/gogo_data_scrubber.cc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        std::endl;
+    return 1;
+  }
+
+  for (int i = argc / 2; i > 0; i--) {
+    const std::string &input_file = argv[i];
+    const std::string &output_file = argv[i + argc / 2];
+
+    std::cerr << "Generating " << input_file <<
+        " to " << output_file << std::endl;
```
Comment (xfxyjwf) [inline benchmarks/util/gogo_data_scrubber.cc]: Fix the format.

File: benchmarks/util/gogo_data_scrubber.cc [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    benchmarks::BenchmarkDataset dataset;
+    Message* message;
+    std::string dataset_payload = ReadFile(input_file);
+    GOOGLE_CHECK(dataset.ParseFromString(dataset_payload))
+    << "Can' t parse data file " << input_file;
```
Comment (xfxyjwf) [inline benchmarks/util/gogo_data_scrubber.cc]: Indent by 4 spaces.

File: benchmarks/util/gogoproto_generator.h [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  if (filename.substr(filename.size() - 11) == ".protodevel") {
+    // .protodevel
+    return filename.substr(0, filename.size() - 11);
+  } else {
+    // .proto
+    return filename.substr(0, filename.size() - 6);
+  }
+}
+
+DescriptorPool new_pool_;
```
Comment (xfxyjwf) [inline benchmarks/util/gogoproto_generator.h]: Please avoid using global objects. Why not make this a member of the GoGoProtoGenerator below?

File: benchmarks/util/gogoproto_generator.h [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  }
+  virtual bool Generate(const FileDescriptor* file,
+                        const string& parameter,
+                        GeneratorContext* context,
+                        string* error) const {
+    if (new_pool_.FindFileByName(file->name())) {
+      return true;
+    }
+    for (int i = 0; i < file->dependency_count(); i++) {
+      Generate(file->dependency(i), parameter, context, error);
```
Comment (xfxyjwf) [inline benchmarks/util/gogoproto_generator.h]: Why are we generating the dependencies here? Can we just generate the file requested?

File: benchmarks/util/gogoproto_generator.h [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    file->CopyTo(&new_file);
+    SchemaGroupStripper::StripFile(file, new_file);
+
+    SchemaAddZeroEnumValue enum_scrubber;
+    enum_scrubber.ScrubFile(new_file);
+
+    string filename = file->name();
+    string basename = StripProto(filename);
+
+    std::vector< std::pair< string, string > > option_pairs;
```
Comment (xfxyjwf) [inline benchmarks/util/gogoproto_generator.h]: Remove the extra spaces around "<" and ">".

File: benchmarks/util/schema_proto2_to_proto3_util.h [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+namespace protobuf {
+namespace util {
+
+class SchemaGroupStripper {
+
+ public:
+  static void StripFile(const FileDescriptor* old_file,
+                                       FileDescriptorProto &file) {
+    for (int i = file.mutable_message_type()->size() - 1; i >= 0; i--) {
+      if (old_file->message_type(i)->options().message_set_wire_format()) {
```
Comment (xfxyjwf) [inline benchmarks/util/schema_proto2_to_proto3_util.h]: Use IsMessageSet() here

File: benchmarks/util/schema_proto2_to_proto3_util.h [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        file.mutable_message_type()->DeleteSubrange(i, 1);
+        continue;
+      }
+      StripMessage(old_file->message_type(i), file.mutable_message_type(i));
+    }
+    for (int i = file.mutable_extension()->size() - 1; i >= 0; i--) {
+      auto field = old_file->extension(i);
+      if (field->type() == FieldDescriptor::TYPE_GROUP
+          || IsMessageSet(field->message_type())
```
Comment (xfxyjwf) [inline benchmarks/util/schema_proto2_to_proto3_util.h]: Put "||" at the end of the line above. Note that we do it differently in C++ and Java.

File: benchmarks/util/schema_proto2_to_proto3_util.h [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+  void ScrubMessage(DescriptorProto *message_type) {
+    for (int i = 0; i < message_type->mutable_enum_type()->size(); i++) {
+      ScrubEnum(message_type->mutable_enum_type(i));
+    }
+    for (int i = 0; i < message_type->mutable_nested_type()->size(); i++) {
+      ScrubMessage(message_type->mutable_nested_type(i));
+    }
+  }
+
+  void ScrubFile(FileDescriptorProto &file) {
```
Comment (xfxyjwf) [inline benchmarks/util/schema_proto2_to_proto3_util.h]: Don't use non-const references in method parameters.
https://google.github.io/styleguide/cppguide.html#Reference_Arguments

File: benchmarks/util/schema_proto2_to_proto3_util.h [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+namespace google {
+namespace protobuf {
+namespace util {
+
+class SchemaGroupStripper {
+
+ public:
+  static void StripFile(const FileDescriptor* old_file,
+                                       FileDescriptorProto &file) {
```
Comment (xfxyjwf) [inline benchmarks/util/schema_proto2_to_proto3_util.h]: Wong indentation.

File: benchmarks/util/schema_proto2_to_proto3_util.h [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+};
+
+class SchemaAddZeroEnumValue {
+
+ public:
+  SchemaAddZeroEnumValue() : total_added_(0) {}
+
+  void ScrubEnum(EnumDescriptorProto *enum_type) {
```
Comment (xfxyjwf) [inline benchmarks/util/schema_proto2_to_proto3_util.h]: Does this need to be public?

File: benchmarks/util/schema_proto2_to_proto3_util.h [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+      }
+      if (!has_zero) {
+        enum_type->mutable_value()->Add();
+        for (int i = enum_type->mutable_value()->size() - 1; i > 0; i--) {
+          enum_type->mutable_value(i)->CopyFrom(
+              *enum_type->mutable_value(i - 1));
+        }
+        enum_type->mutable_value(0)->set_number(0);
+        std::stringstream ss;
+        ss << "ADDED_ZERO_VALUE_" << total_added_++;
```
Comment (xfxyjwf) [inline benchmarks/util/schema_proto2_to_proto3_util.h]: You can use std::to_string(int) in C++11.

File: benchmarks/util/schema_proto2_to_proto3_util.h (hunk 15 above)
Comment (BSBandme) [reply benchmarks/util/schema_proto2_to_proto3_util.h]: Done

File: benchmarks/util/schema_proto2_to_proto3_util.h (hunk 18 above)
Comment (BSBandme) [reply benchmarks/util/schema_proto2_to_proto3_util.h]: Done

File: benchmarks/util/schema_proto2_to_proto3_util.h (hunk 17 above)
Comment (BSBandme) [reply benchmarks/util/schema_proto2_to_proto3_util.h]: Done by moving to private

File: benchmarks/util/schema_proto2_to_proto3_util.h (hunk 14 above)
Comment (BSBandme) [reply benchmarks/util/schema_proto2_to_proto3_util.h]: Done

File: benchmarks/util/schema_proto2_to_proto3_util.h (hunk 13 above)
Comment (BSBandme) [reply benchmarks/util/schema_proto2_to_proto3_util.h]: Done

File: benchmarks/util/schema_proto2_to_proto3_util.h (hunk 16 above)
Comment (BSBandme) [reply benchmarks/util/schema_proto2_to_proto3_util.h]: Done

File: benchmarks/util/gogoproto_generator.h (hunk 12 above)
Comment (BSBandme) [reply benchmarks/util/gogoproto_generator.h]: Done

File: benchmarks/util/gogoproto_generator.h (hunk 11 above)
Comment (BSBandme) [reply benchmarks/util/gogoproto_generator.h]: Change deps check to GenerateAll

File: benchmarks/util/gogoproto_generator.h (hunk 10 above)
Comment (BSBandme) [reply benchmarks/util/gogoproto_generator.h]: The Generate function need "this" to be const. I'm not sure in which way I can avoid global objects.

File: benchmarks/util/gogoproto_generator.h (hunk 6 above)
Comment (BSBandme) [reply benchmarks/util/gogoproto_generator.h]: Done

File: benchmarks/util/gogo_data_scrubber.cc (hunk 9 above)
Comment (BSBandme) [reply benchmarks/util/gogo_data_scrubber.cc]: Done

File: benchmarks/util/gogo_data_scrubber.cc (hunk 8 above)
Comment (BSBandme) [reply benchmarks/util/gogo_data_scrubber.cc]: Done

File: benchmarks/util/gogo_data_scrubber.cc (hunk 7 above)
Comment (BSBandme) [reply benchmarks/util/gogo_data_scrubber.cc]: Done

File: benchmarks/gogo/go_no_group/go_benchmark_test.go (hunk 4 above)
Comment (BSBandme) [reply benchmarks/gogo/go_no_group/go_benchmark_test.go]: Done by add rule generate_all_gogo_benchmark_code

File: benchmarks/gogo/cpp_no_group/cpp_benchmark.cc (hunk 3 above)
Comment (BSBandme) [reply benchmarks/gogo/cpp_no_group/cpp_benchmark.cc]: Done by add rule generate_cpp_no_group_benchmark_code

File: benchmarks/Makefile.am [hunk 19]
```diff
@@ -253,50 +252,313 @@ benchmarks_protoc_inputs_proto2_message4 =                                 \
...
+	mkdir -p tmp/go_no_group
+	mkdir -p tmp/gogofast
+	mkdir -p tmp/gogofaster
+	mkdir -p tmp/gogoslick
+	touch make_tmp_dir_gogo
+
+if USE_EXTERNAL_PROTOC
+
+go_no_group_protoc_middleman: make_tmp_dir_gogo gogo_proto_middleman
+	cd $(srcdir)/tmp/gogo_proto && $(PROTOC) -I$(srcdir) -I$(top_srcdir) --go_out=$$oldpwd/tmp/go_no_group $(benchmarks_protoc_inputs)
```
[the remark this answers is not in the document]
Comment (BSBandme) [reply benchmarks/Makefile.am]: May be possible for cpp, but for go it only can generate proto code with the same package once.

File: benchmarks/Makefile.am (hunk 1 above)
Comment (BSBandme) [reply benchmarks/Makefile.am]: Do you mean add some variable during configure?

File: benchmarks/Makefile.am [hunk 20]
```diff
@@ -253,50 +252,313 @@ benchmarks_protoc_inputs_proto2_message4 =                                 \
...
+	$(cpp_no_group_benchmarks_protoc_outputs_proto2_header) $(top_srcdir)/third_party/benchmark/src/libbenchmark.a
+nodist_cpp_no_group_benchmark_SOURCES =                                    \
+	$(cpp_no_group_benchmarks_protoc_outputs_proto2)                         \
+	$(cpp_no_group_benchmarks_protoc_outputs)                                \
+	$(cpp_no_group_benchmarks_protoc_outputs_header)                         \
+	$(cpp_no_group_benchmarks_protoc_outputs_proto2_header)
+ 
+gogo_proto_middleman: protoc-gen-gogoproto
+	mkdir -p "tmp/gogo_proto"
+if USE_EXTERNAL_PROTOC
```
[the remark this answers is not in the document]
Comment (BSBandme) [reply benchmarks/Makefile.am]: I'm not sure about that. This is previously added, if you think this is not necessary I'll delete it

File: benchmarks/Makefile.am [hunk 21]
```diff
@@ -253,50 +252,305 @@ benchmarks_protoc_inputs_proto2_message4 =                                 \
...
+	touch generate_gogo_data
+	
+make_tmp_dir_gogo:
+	mkdir -p tmp/go_no_group/benchmark_code
+	mkdir -p tmp/gogofast/benchmark_code
+	mkdir -p tmp/gogofaster/benchmark_code
+	mkdir -p tmp/gogoslick/benchmark_code
+	touch make_tmp_dir_gogo
+
+if USE_EXTERNAL_PROTOC
```
Comment (xfxyjwf) [inline benchmarks/Makefile.am]: Let's remove the use of "USE_EXTERNAL_PROTOC". It's used for cross-compile but we don't use it for benchmarks.

**Pull request merged**